### ISSUE TITLE
Server FPS drops caused by masterIdTracker

### ISSUE THREAD

**Issue by tbeswick96** [IDI-Systems/acre2#54]
**Arma 3 Version:** `1.64 & 1.62` stable
**CBA Version:** `3.0.0` stable (latest)
**ACRE2 Version:** `2.3.0.905/6` dev https://github.com/IDI-Systems/acre2/commit/c115377a3067aaef8cbf6b26f628fc8032ea41bf

**Mods:**
- `@CBA_A3`
- `@acre2`

**Description:**
- Severe server FPS drops every 10 seconds.
- I have thoroughly tested this, and believe I've found the cause. I will go through my entire findings to paint a clearer picture.
- Last Saturday, we (the unit I play with, 30-40 guys) started playing an op, with at least 300+ ai. I noticed some pretty severe desync, and monitoring server FPS saw a consistent drop to 1 from ~45 every second. Post-op, I did some further testing by removing mods, changing versions etc. I found the drops to be caused by ACRE.
- Today, I cloned this git, and tested with the build from the latest commit. I no longer had the server FPS drop every second. However, I noticed a new drop every 10 seconds. Same as before, average 45 server FPS, dropping to <20 every 10 seconds, depending on the number of ai/vehicles active at the time.
- Through some poking, I found the masterIdTracker to be the cause, as this was the only thing I could find that runs every 10 seconds on the server. I commented out line 21 in  
  `addons/sys_server/fnc_masterIdTracker.sqf: GVAR(doFullSearch) = true;`  
  so it no longer did the full search every 10 seconds. I built and tested again, with the same mission file, and the server FPS drop every 10 seconds had gone.
- I believe this to be the line that causes the FPS drop, as it scales with the number of units, players, vehicles, containers active in the mission:  
  `private _searchObjects = allPlayers + allUnits + allDead + vehicles + (allMissionObjects "WeaponHolder"); // search players first`  
- I don't know the full repercussions of stopping this check, but they cannot be good if the check needs to be done every 10 seconds.
- This is a fairly severe issue in my opinion, as units with larger numbers of players and big missions get hit hard by desync, making it fairly unplayable.

**Steps to reproduce:**
- Load only CBA and ACRE built from latest commit
- Make a mission with 1 player, and a gamemaster module so you can use zeus.
- Launch a dedicated server with the same mods enabled.
- Join the server and login as admin. Run `#monitor 0.1`
- Watch the server FPS. With just 1 player placed, this should be a fairly constant ~45 FPS.
- Place a group of ai and give them a waypoint so they're doing something. Copy and paste the group with the waypoint enough times to have at least 200 ai active in the scenario.
- Wait for the FPS to settle back to ~45.
- Look carefully and you should see FPS drops every 10 seconds. Do `#monitor 0.2` if it is too fast.
- To exacerbate the issue, place more ai or anything that could be a container.

**Where did the issue occur?**
- Dedicated

**Placed Modules:**
- None.

**RPT log file:**
- Honestly too many logs to count, but I doubt what you'd need would be in logs anyway.

**Comment by tbeswick96** [comment]
A comment on https://community.bistudio.com/wiki/allMissionObjects 

`
Be VERY careful with the use of this command. It is very demanding as it must iterate through all mission created objects. Particular care should be taken exercising this often on dedicated servers.
`

**Comment by Sniperhid** [comment]
We are aware of this design. Note this is the most reliable method to reliably ensure all containers are checked at once and thus all radios are discovered and checked by ACRE.

Whilst the FPS 'drop' still exists compared to last stable release some significant improvements were made
#39 - Should reduce the load by 50% on average (maybe more/less in your case).
And yes compared to last stable the checks happen less frequently (every 10 seconds versus 1 second).

This pattern has existed in ACRE2 for a longtime and I have played many large coop missions without issue, even if server FPS periodically takes a hit. From my own experience 300 AI without caching/headless client can cause quite a server load without ACRE2.

If someone wants to propose a better implementation by all means, this is largely an old system that could possibly benefit with a re-design.

**Comment by kami-** [comment]
How about have 2 configurable options:
1. Leave the current system running by default.
2. Have an option to disable current tracking loop to discover radios. Instead, use dedicated ACRE API functions to add radios to units/vehicles. This would let power users to manage the radios manually and improve performance. Would untracked radios break ACRE?

**Comment by jonpas** [comment]
@kami- No, as soon as we add something like that we will definitely get reports of issues which that scan is required by to work properly.

For now we (ACRE2Team) don't have plans on improving it, I tagged the issue as possible optimization. Maybe once we announce this GitHub publicly someone will take up the task, however it's no easy task.

**Comment by tbeswick96** [comment]
Wrapping `doFullSearch` in a spawn completely removes the FPS drops

```
if(GVAR(doFullSearch)) then {
    [] spawn {
        ...
    };
};
```

**Comment by Sniperhid** [comment]
And radios can move locations between frames thus it checks at a given instant (single frame). If someone was in the process of dropping/picking up a radio it would be possible for the fullsearch to simply not find the item and assume it had been deleted resulting in its data being garbage collected. So using a spawn removes the reliability of the check.

**Comment by tbeswick96** [comment]
I reverted my changes back to how it was since the latest commit for masterIdTracker, and we ran the same mission on Saturday.
There were no desync issues relting to this, so I think the optimisation you guys did already has solved the issue.
The spikes every 10 seconds still persisted, however it did not cause any issues at all. This was with the mission with roughly 400 ai placed, with roughly 150 maximum active and simulated at any one time thanks to caching.

All in all, I don't think you need worry about this issue :)

**Comment by Soldia1138** [comment]
Thank you for the feedback

**Comment by NouberNou** [comment]
Close?

**Comment by PabstMirror** [comment]
Not perfect accuracy, but a good visual of the change from 904 to current master:

904 - http://i.imgur.com/asS5roQ.jpg
- Max is about 25ms over normal every second

Current - http://i.imgur.com/hRiL43f.jpg
- Max is about 17ms over normal every 10 second

So a roughly 15x speedup.

You could possibly split the scan over multiple frames, and only garbage collect the ID if it's been missing for multiple scans. But doing it atomically is the safest and simplest way to get it done.

**Comment by jonpas** [comment]
Yeah multiple frame is a problem as things can change during those frames, what you proposed is something @snippers was thinking as well. It's quiet a good improvement he did already, I'd say close this issue and further improvements can be done at any later time.

**Comment by rebelvg** [comment]
As a mission maker you can remove items from ai units and if you have a lot of items in your boxes, make them have "infinite" number of items, constantly re-adding items that were taken by players. Or you can make the system that re-adds items and virtually counting how many were taken, so that box doesn't really have infinite items.
Here's an example.
https://github.com/rebelvg/framework.VR/blob/cca33337799cc5e43fca7b167b98b176835bf4fb/framework/functions/fn_frameworkInit.sqf#L188-L239
This makes items count to 2 items and re-adds as soon as item was taken. Basically it makes stuff like bandages infinite which is not a really big problem, mags are not touched by this.

**Comment by Sniperhid** [comment]
Further optimisation and speedups to the master ID tracker were implemented in #179

On a side note #180 offers some flexibility to radios that would traditionally be garbage collected.
